src/SFML/Window/Win32/JoystickImpl.cpp

```diff
@@ -101,6 +117,215 @@ JoystickBlacklist joystickBlacklist;
 const DWORD directInputEventBufferSize = 32;
 } // namespace
 
+// for XInput
+namespace
+{
+constexpr std::size_t xinputMaxDevices{4};
+std::vector<bool>     xInputSlots{};
+bool                  directInputNeedsInvalidation{};
+
+[[nodiscard]] sf::Joystick::Axis getAxis(std::size_t index)
+{
+    switch (index)
+    {
+        case 0:
+            return sf::Joystick::Axis::X;
+        case 1:
+            return sf::Joystick::Axis::Y;
+        case 2:
+            return sf::Joystick::Axis::Z;
+        case 3:
+            return sf::Joystick::Axis::R;
+        case 4:
+            return sf::Joystick::Axis::U;
+        case 5:
+            return sf::Joystick::Axis::V;
+        case 6:
+            return sf::Joystick::Axis::PovX;
+        case 7:
+            return sf::Joystick::Axis::PovY;
+        default:
+            throw std::out_of_range("index is out of range.");
+    }
+}
+
+struct XInputCleanupData
+{
+    VARIANT                           var            = {};
+    IWbemLocator*                     pIWbemLocator  = nullptr;
+    IEnumWbemClassObject*             pEnumDevices   = nullptr;
+    std::array<IWbemClassObject*, 20> pDevices       = {};
+    IWbemServices*                    pIWbemServices = nullptr;
+    BSTR                              bstrNamespace  = nullptr;
+    BSTR                              bstrDeviceID   = nullptr;
+    BSTR                              bstrClassName  = nullptr;
+};
+
+// NOLINTBEGIN(readability-identifier-naming)
+// keeping GUIDs and UUIDs consistent
+// Define CLSID_WbemLocator
+const CLSID CLSID_WbemLocator = {0x4590f811, 0x1d3a, 0x11d0, {0x89, 0x1f, 0x00, 0xaa, 0x00, 0x4b, 0x2e, 0x24}};
+
+// Define IID_IWbemLocator
+const IID IID_IWbemLocator = {0xdc12a687, 0x737f, 0x11cf, {0x88, 0x4d, 0x00, 0xaa, 0x00, 0x4b, 0x2e, 0x24}};
+// NOLINTEND(readability-identifier-naming)
+
+// Function pointer type for XInputGetState
+using XInputGetState_t = DWORD(WINAPI*)(DWORD dwUserIndex, XINPUT_STATE* pState);
+
+XInputGetState_t mXInputGetState = nullptr;
+
+void safeCleanup(XInputCleanupData& data)
+{
+    VariantClear(&data.var);
+
+    if (data.bstrNamespace)
+        SysFreeString(data.bstrNamespace);
+    if (data.bstrDeviceID)
+        SysFreeString(data.bstrDeviceID);
+    if (data.bstrClassName)
+        SysFreeString(data.bstrClassName);
+
+    for (auto* device : data.pDevices)
+        SafeRelease(device);
+
+    SafeRelease(data.pEnumDevices);
+    SafeRelease(data.pIWbemLocator);
+    SafeRelease(data.pIWbemServices);
+}
+
+// See also https://learn.microsoft.com/en-us/windows/win32/xinput/xinput-and-directinput?redirectedfrom=MSDN
+[[nodiscard]] BOOL isXInputDevice(const GUID* pGuidProductFromDirectInput)
+{
+    XInputCleanupData data;
+    bool              bisXInputDevice = false;
```

Comment:
avoid C style Hungarian naming the `is` clearly tells us it is a boolean.


```suggestion
    bool              isXInputDevice = false;
```

Reply:
Good idea, that variable name was a holdover from the MSFT example that I copied over. Implemented.